### DISCUSSION TITLE
What is the purpose of Docker DinD?

### DISCUSSION THREAD

**Discussion by AmadeoRR** [community/community#160725] [Actions]
Hello, so we are currently going through the process of procuring more information on this actions runner controller and would like to get some more information from the team managing it. 

What was the purpose of using Docker DinD in this app? And is it required that you use some sort of privileged containers for the image? Basically we are curious why it was needed in this case as we would assume dependency management could be handled without root or privileged containers from the onset. 

WHat is the purpose of Docker DinD in the first place? Thanks

**Comment by cbui** [comment]
ARC itself runs as a container, so Docker DinD is needed to support the use case where users have workflows that run in docker containers. 

> And is it required that you use some sort of privileged containers for the image?

Yes.

**Reply by shichengripple001** [reply]
can't we just mount docker socket to runner container? to save resource

**Reply by cbui** [reply]
You can with this setting: `dockerdWithinRunnerContainer`

**Reply by shichengripple001** [reply]
thx. will try

**Reply by roshvin** [reply]
hi @shichengripple001  are you able to get DID and build docker container from your runner . we are also looking for this option and not finding a solution.

**Reply by usernamessh** [reply]
> ARC itself runs as a container, so Docker DinD is needed to support the use case where users have workflows that run in docker containers.

Picking up this topic again:
Can you please explain, how to run such workflows in this DinD container then? I mean, the DinD container aparently isn't started with "docker run -it -d <image>", so it just terminates soon after being started. So how to interact with it from workflow level e. g. copying files into into it?

**Comment by mumoshu** [comment]
Hey. `privileged` is required to run privileged operations from within jobs (like bringing up and down VMs to run some job steps) or for `dind`. If you don't need `dind` and other privileged operations, ARC 0.25.0 or greater has support for actions/runner's new "kubernetes" container mode, which allows you to ditch `privileged`.
https://github.com/actions-runner-controller/actions-runner-controller/blob/master/docs/releasenotes/0.25.md#enhancement--support-for-the-new-kubernetes-container-mode-of-actions-runner
